core/src/main/java/org/infinispan/functional/impl/EntryViews.java

```diff
@@ -32,81 +35,106 @@ private EntryViews() {
+      return new NoValueReadOnlyView<>(key, cacheEncoders);
    }
 
    /**
     * For convenience, a lambda might decide to return the entry view it
     * received as parameter, because that makes easy to return both value and
     * meta parameters back to the client.
-    *
+    * <p>
     * If the lambda function decides to return an view, launder it into an
     * immutable view to avoid the user trying apply any modifications to the
     * entry view from outside the lambda function.
-    *
+    * <p>
     * If the view is read-only, capture its data into a snapshot from the
     * cached entry and avoid changing underneath.
     */
    @SuppressWarnings("unchecked")
-   public static <R> R snapshot(R ret) {
+   public static <R> R snapshot(R ret, CacheEncoders cacheEncoders) {
       if (ret instanceof EntryBackedReadWriteView) {
          EntryBackedReadWriteView view = (EntryBackedReadWriteView) ret;
-         return (R) new ReadWriteSnapshotView(view.key(), view.entry.getValue(), view.entry.getMetadata());
+         return (R) new ReadWriteSnapshotView(view.key(), view.find().orElse(null), view.entry.getMetadata());
       } else if (ret instanceof EntryAndPreviousReadWriteView) {
          EntryAndPreviousReadWriteView view = (EntryAndPreviousReadWriteView) ret;
-         return (R) new ReadWriteSnapshotView(view.key(), view.entry.getValue(), view.entry.getMetadata());
+         return (R) new ReadWriteSnapshotView(view.key(), view.getCurrentValue(), view.entry.getMetadata());
       } else if (ret instanceof EntryBackedReadOnlyView) {
          EntryBackedReadOnlyView view = (EntryBackedReadOnlyView) ret;
-         return (R) new ReadOnlySnapshotView(view.key(), view.entry.getValue(), view.entry.getMetadata());
+         return (R) new ReadOnlySnapshotView(view.key(), view.find().orElse(null), view.entry.getMetadata());
+      } else if (ret instanceof NoValueReadOnlyView) {
+         NoValueReadOnlyView view = (NoValueReadOnlyView) ret;
+         return (R) new ReadOnlySnapshotView(view.key(), null, null);
       }
 
-      return ret;
+      return (R) cacheEncoders.valueFromStorage(ret);
    }
 
    private static final class EntryBackedReadOnlyView<K, V> implements ReadEntryView<K, V> {
       final CacheEntry<K, V> entry;
+      private final CacheEncoders cacheEncoders;
 
-      private EntryBackedReadOnlyView(CacheEntry<K, V> entry) {
+      private EntryBackedReadOnlyView(CacheEntry<K, V> entry, CacheEncoders cacheEncoders) {
          this.entry = entry;
+         this.cacheEncoders = cacheEncoders;
       }
 
       @Override
       public K key() {
-         return entry.getKey();
+         return (K) cacheEncoders.keyFromStorage(entry.getKey());
       }
 
       @Override
       public Optional<V> find() {
-         return entry == null ? Optional.empty() : Optional.ofNullable(entry.getValue());
+         return entry == null ? Optional.empty() : Optional.ofNullable((V) cacheEncoders.valueFromStorage(entry.getValue()));
```

Comment:
@rvansa you mean that if find() is called several times, to have the decoded value already decoded and available ?